src/node.h

```diff
@@ -251,7 +251,13 @@ inline void NODE_SET_METHOD(const TypeName& recv,
   v8::Local<v8::Function> fn = t->GetFunction();
   v8::Local<v8::String> fn_name = v8::String::NewFromUtf8(isolate, name);
   fn->SetName(fn_name);
-  recv->Set(fn_name, fn);
+  auto getter = [](v8::Local<v8::String>,
+                   const v8::PropertyCallbackInfo<v8::Value>& info) {
+    auto data = info.Data();
```

Comment:
this doesn't really fix the issue.

Sorry if my initial description of the problem and the solution wasn't clear. I assume that the security aspect of cross-context leaks aren't much of a concern to node as node doesn't use contexts to separate different scripts that don't trust each other. The main concern here is memory leaks.

Templates are per-isolate objects, while non-primitive values are per context objects. By directly (or indirectly) putting a  per-context value on a per-isolate value, you essentially disable GC for all objects in that context (as every non-primitive value keeps its context alive which in turn keeps any object in that context alive).

You might expect that instantiating a template creates a copy of all the values, but it does not, it just reuse the original value you've put in (as it's generally not possible to clean objects - think internal fields: the VM doesn't know whether it's safe to duplicate C pointers).

here, you already have a template for fn, you should can just recv->Set(fn_name, t);

when recv is instantiated, a function in the target context will be created using the t template.